cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -449,6 +449,63 @@ class TimestampConverter
   TimeUnit::type unit_;
 };
 
+template <NullCoding null_coding>
+class DurationConverter
```

Comment:
This is almost literally a copy paste of the TimestampConverter above with timestamp-things replaced with duration-things. Is there a way to deduplicate some code? 
Templating this might be difficult as I also need to switch Python / numpy C API calls? (`PyDateTime_DateTime` -> `PyDateTime_Delta`, `PyDatetimeScalarObject` -> `PyTimedeltaScalarObject`, ...)

Reply:
Can use something like this:
```cpp
template <typename ArrowType>
struct PyDateTimeTraits {};

template <>
struct PyDateTimeTraits<TimestampType> {
  using PyTypeCheck = PyDateTime_Check;
  using PyDateTimeObject = PyDateTime_DateTime;

  using py_to_s = internal::PyDateTime_to_s;
  using py_to_ms = internal::PyDateTime_to_ms;
  using py_to_us = internal::PyDateTime_to_us;
  using py_to_ns = internal::PyDateTime_to_ns;

  using np_traits = internal::npy_traits<NPY_DATETIME>;
  using NpScalarObject = PyDatetimeScalarObject;
};
```

Reply:
It seems that `using` can only be used with type names, and not function names? 
(I could define an inline function that wraps the correct function for each, but that gets a bit verbose?)

Reply:
Inline functions sound fine.

Reply:
OK, I pushed a version with the template (I am personally not fully sure the end result is easier / better than some duplicated code, but ok ..)